shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java

```diff
@@ -108,72 +108,77 @@ public Counter call() throws Exception {
     }
 
     @Override
-    public void updateCounterValue(String name, long value) {
+    public void updateCounterValue(String name, long value, String... tags) {
         Exceptions.checkNotNullOrEmpty(name, "name");
-        String counterName = name + ".Counter";
+        String counterName = nameFromTags(name, tags) + ".Counter";
 
         Counter counter = countersCache.getIfPresent(counterName);
         if (counter != null) {
             counter.clear();
         } else {
-            counter = underlying.createCounter(counterName);
+            counter = underlying.createCounter(
+                    (tags == null || tags.length == 0) ? counterName : name, tags);
         }
         counter.add(value);
         countersCache.put(name, counter);
     }
 
     @Override
-    public void freezeCounter(String name) {
-        String counterName = name + ".Counter";
+    public void freezeCounter(String name, String... tags) {
+        String counterName = nameFromTags(name, tags) + ".Counter";
+        Counter counter = countersCache.getIfPresent(counterName);
+        metrics.remove(counter.getId());
         countersCache.invalidate(counterName);
-        metrics.remove(counterName);
     }
 
     @Override
-    public <T extends Number> void reportGaugeValue(String name, T value) {
+    public <T extends Number> void reportGaugeValue(String name, T value, String... tags) {
         Exceptions.checkNotNullOrEmpty(name, "name");
         Preconditions.checkNotNull(value);
         Gauge newGauge = null;
-        String gaugeName = name + ".Gauge";
+        String gaugeCacheKey = nameFromTags(name, tags) + ".Gauge";
+        String gaugeRegisterKey = (tags == null || tags.length == 0) ? gaugeCacheKey : name;
```

Comment:
This conditional pattern is used throughout this file. Can you abstract it out?

Reply:
sure